### PR TITLE
Stats collection

### PR DESCRIPTION
**Pull request by real-danm**
Requires https://github.com/livekit/protocol/pull/263

### CODE REVIEW

File: deploy/grafana/livekit-server-overview.json [hunk 1]
```diff
@@ -598,7 +598,7 @@
         },
         {
           "exemplar": true,
-          "expr": "sum(rate(livekit_node_service_operation{status=\"success\", type=\"signal_ws\"}[5m])) - sum(rate(livekit_node_service_operation{status=\"success\", type=\"ice_connection\"}[5m]))",
+          "expr": "sum(rate(livekit_node_service_operation{status=\"success\", type=\"signal_ws\"}[5m])) - (sum(rate(livekit_node_service_operation{status=\"success\", type=\"ice_connection\"}[5m]))/2)",
```
Comment (real-danm) [inline deploy/grafana/livekit-server-overview.json:601]: We create a publisher and subscriber per participant in transportmanager so there are 2 ice_connection to 1 signal_ws

**Review by matkam — APPROVED** [review]
lgtm!

File: pkg/rtc/mediatrack.go [hunk 2]
```diff
@@ -196,13 +197,15 @@ func (t *MediaTrack) AddReceiver(receiver *webrtc.RTPReceiver, track *webrtc.Tra
...
 		t.params.Logger.Errorw("could not retrieve buffer pair", nil)
+		prometheus.AddPublishedTrackFailure(t.trackInfo.Type.String())
 		return newCodec
 	}
 
 	rtcpReader.OnPacket(func(bytes []byte) {
 		pkts, err := rtcp.Unmarshal(bytes)
 		if err != nil {
 			t.params.Logger.Errorw("could not unmarshal RTCP", err)
+			prometheus.AddPublishedTrackFailure(t.trackInfo.Type.String())
```
Comment (davidzhao) [inline pkg/rtc/mediatrack.go:208]: this is not the right place for it. this is called on each packet. Also, I think publishing checks can be performed by the supervisor framework.

Another issue with keeping track of failures: we don't always know when something has failed (but will know that they've succeeded). For example, if the onTrack function doesn't fire, we would not be able to account for that.

Therefore, I think in general it's much more predictable to keep track of successes and total. failure rate could be derived from that.

File: pkg/rtc/room.go [hunk 3]
```diff
@@ -221,17 +221,17 @@ func (r *Room) Join(participant types.LocalParticipant, opts *ParticipantOptions
 	defer r.lock.Unlock()
 
 	if r.IsClosed() {
-		prometheus.ServiceOperationCounter.WithLabelValues("participant_join", "error", "room_closed").Add(1)
```
Comment (davidzhao) [inline pkg/rtc/room.go:224]: I'm not sure that `participant_join` or error rate is actually a useful metric here. They are rare, and they are not really actionable.

I think the piece we really focus on is participants connecting to the server.

Comment (davidzhao) [reply pkg/rtc/room.go:224]: should we remove this? wdyt @boks1971 @matkam @cnderrauber

File: pkg/rtc/transport.go [hunk 4]
```diff
@@ -1491,7 +1491,7 @@ func (t *PCTransport) createAndSendOffer(options *webrtc.OfferOptions) error {
 
 	offer, err := t.pc.CreateOffer(options)
 	if err != nil {
-		prometheus.ServiceOperationCounter.WithLabelValues("offer", "error", "create").Add(1)
+		prometheus.AddServiceOperation("offer", "error", "create")
```
Comment (davidzhao) [inline pkg/rtc/transport.go:1494]: I think I've never/rarely seen this fail either.

File: pkg/telemetry/events.go [hunk 5]
```diff
@@ -68,7 +68,6 @@ func (t *telemetryService) ParticipantJoined(
 	shouldSendEvent bool,
 ) {
 	t.enqueue(func() {
-		prometheus.IncrementParticipantJoin(1)
```
Comment (davidzhao) [inline pkg/telemetry/events.go:71]: was this duplicated?

File: pkg/telemetry/prometheus/node.go [hunk 6]
```diff
@@ -97,6 +101,37 @@ func getMemoryStats() (memoryLoad float32, err error) {
 	return
 }
 
+func AddServiceOperation(typeStr string, status string, errType string) {
```
Comment (davidzhao) [inline pkg/telemetry/prometheus/node.go:104]: we should use an aliased type vs strings.

File: pkg/telemetry/prometheus/node.go [hunk 7]
```diff
@@ -97,6 +101,37 @@ func getMemoryStats() (memoryLoad float32, err error) {
...
+	switch typeStr {
+	case "ice_connection", "peer_connection", "offer", "answer":
+		if errType == "" {
+			connectionSuccess.Inc()
```
Comment (davidzhao) [inline pkg/telemetry/prometheus/node.go:108]: what is the goal we are trying to measure here? if it's a participant connection, then recording these would be drastically over-counting things.

File: pkg/telemetry/prometheus/rooms.go [hunk 8]
```diff
@@ -106,3 +112,7 @@ func SubSubscribedTrack(kind string) {
 	promTrackSubscribedTotal.WithLabelValues(kind).Sub(1)
 	trackSubscribedTotal.Dec()
 }
+
+func AddSubscribedTrackFailure(kind string) {
```
Comment (davidzhao) [inline pkg/telemetry/prometheus/rooms.go:116]: where is this used?

File: pkg/rtc/room.go (hunk 3 above)
Comment (boks1971) [reply pkg/rtc/room.go:224]: Yeah, I think we can remove all the counters here. These are not problems per se. These are conditions under which participant should not be allowed to join. Maybe logs here, but not stats counters.

File: pkg/rtc/room.go [hunk 9]
```diff
@@ -323,7 +323,7 @@ func (r *Room) Join(participant types.LocalParticipant, opts *ParticipantOptions
 
 	joinResponse := r.createJoinResponseLocked(participant, iceServers)
 	if err := participant.SendJoinResponse(joinResponse); err != nil {
-		prometheus.ServiceOperationCounter.WithLabelValues("participant_join", "error", "send_response").Add(1)
+		prometheus.AddServiceOperation("participant_join", "error", "send_response")
```
Comment (boks1971) [inline pkg/rtc/room.go:326]: Maybe, this counter can be there.

**Comment by real-danm** [comment]
Reworking this.

**Pull request closed without merging**